### PR TITLE
Ability to load font from remote uri

### PR DESCRIPTION
**Pull request by seekshiva**
# Why

Calling `Font.loadAsync` with a remote uri currently throws an error saying that remote uri file download isn't supported yet. This is my initial attempt at implementing `Asset.fromUri`, thereby allowing remote uri fonts to be loaded to the app.

# How

This contains a naive implementation for `fromUri`. It is also not possible to know the md5 hash of the file beforehand, like the way we know with other normal type assets. So I've added a constructor argument for Asset called `ignoreHashCheck` which lets us ignore hash check for remote uri assets.

The implementation in this PR makes remote font loading for my specific use case, but if someone has suggestion for extending this to more generic use cases, please let me know.

**Caveat:** It is not possible to know the md5 hash of files beforehand, so this PR ignores the hash check. This also means that newer version of files would not be re-downloaded if the file already exists in cache directory.

The name of the file that the remote file gets downloaded to is the base64 encoded value of the uri itself. This ensures that different uri have different their own unique local file name and wouldn't conflict. Is there a better way to do this?

Also, prettier auto-formatted a bunch of other lines in the Asset file that are not related to the change. Please let me know if I should reset them or move that change to a separate commit.

# Test Plan

```
Expo.Font.loadAsync({
  Montserrat: "https://host.seekshiva.in/downloads/fonts-master/ofl/montserrat/Montserrat-Regular.ttf",
  'Montserrat-SemiBold': "https://host.seekshiva.in/downloads/fonts-master/ofl/montserrat/Montserrat-SemiBold.ttf",
});
```

### CODE REVIEW

File: packages/expo-asset/src/Asset.js [hunk 1]
```diff
@@ -193,28 +212,48 @@ export default class Asset {
     return asset;
   }
 
+  static fromUri(remoteUri) {
+    const hash = Buffer.from(remoteUri).toString('base64');
+    const pathArr = remoteUri.split('?')[0].split('/');
```
Comment (ide) [inline packages/expo-asset/src/Asset.js]: Use uriparser instead (already used here)

File: packages/expo-asset/src/Asset.js [hunk 2]
```diff
@@ -193,28 +212,48 @@ export default class Asset {
...
+    const filename = pathArr[pathArr.length - 1];
+    const [name, ext] = filename.split('.');
+    const arr = remoteUri.split('.');
+    const asset = new Asset({
+      name,
+      uri: remoteUri,
+      hash,
```
Comment (ide) [inline packages/expo-asset/src/Asset.js]: Pass in null if there's no hash

File: packages/expo-asset/src/Asset.js [hunk 3]
```diff
@@ -193,28 +212,48 @@ export default class Asset {
...
+      const localUri = `${FS.cacheDirectory}ExponentAsset-${this.hash}.${
+        this.type
+      }`;
       let exists, md5;
       ({ exists, md5 } = await FS.getInfoAsync(localUri, {
         cache: true,
         md5: true,
       }));
-      if (!exists || md5 !== this.hash) {
+      if (!exists || (!this.ignoreHashCheck && md5 !== this.hash)) {
```
Comment (ide) [inline packages/expo-asset/src/Asset.js]: We usually want to check the hash and skip it just when we have no hash:

```suggestion
      if (!exists || (this.hash !== null && md5 !== this.hash)) {
```

**Review by ide — CHANGES_REQUESTED** [review]

**Comment by ide** [comment]
Please also add a small demo to the native-component-list app and a test to test-suite, both under `apps`.

File: packages/expo-asset/src/Asset.js [hunk 4]
```diff
@@ -193,28 +212,48 @@ export default class Asset {
...
     if (this.downloading) {
-      await new Promise((resolve, reject) => this.downloadCallbacks.push({ resolve, reject }));
+      await new Promise((resolve, reject) =>
+        this.downloadCallbacks.push({ resolve, reject })
+      );
       return;
     }
     this.downloading = true;
     try {
-      const localUri = `${FS.cacheDirectory}ExponentAsset-${this.hash}.${this.type}`;
```
Comment (seekshiva) [inline packages/expo-asset/src/Asset.js]: @ide How should we construct the localUri if hash is null?

**Comment by seekshiva** [comment]
All changes mentioned above are done. Adding test to test-suite is pending.

**Comment by ide** [comment]
Waiting for test suite and a rebase before review again

**Comment by seekshiva** [comment]
I've redone the changes on top of .ts files and pushed. Can you help me with running the test suite locally, I get this error while trying to run `yarn test` inside expo-assets folder:

```
  ● Test suite failed to run

    Cannot find module 'setupDevtools' from 'setup.js'

      at Resolver.resolveModule (../../node_modules/jest-resolve/build/index.js:221:17)
      at Object.<anonymous> (../../node_modules/react-native/jest/setup.js:3:6)
```

**Comment by seekshiva** [comment]
@ide Please let me know how we can proceed forward with this change.

**Comment by ide** [comment]
CI (the sdk workflow) seems to be passing so if you can't get things working locally (probably wrong dep versions somewhere) you can look at CI.

**Comment by sjchmiela** [comment]
I somehow failed to keep the original ownership of the changes when merging an updated branch, sorry! Your contribution is still listed in the `CHANGELOG.md`. Thanks!

**Comment by seekshiva** [comment]
Thank you so much

**Pull request closed without merging**